test/solvers/advanced/forward_backward_test.py

```diff
@@ -0,0 +1,227 @@
+
+# External
+import pytest
+import numpy as np
+
+# Internal
+import odl
+from odl.solvers import forward_backward_pd
+from odl.util.testutils import all_almost_equal, almost_equal, example_element
+
+# Places for the accepted error when comparing results
+HIGH_ACCURACY = 8
+LOW_ACCURACY = 4
+
+
+def test_forward_backward_input_handling():
+    """Test to see that input is handled correctly."""
+
+    space1 = odl.uniform_discr(0, 1, 10)
+
+    lin_ops = [odl.ZeroOperator(space1), odl.ZeroOperator(space1)]
+    prox_cc_g = [odl.solvers.proximal_zero(space1),  # Identity operator
+                 odl.solvers.proximal_zero(space1)]  # Identity operator
+    prox_f = odl.solvers.proximal_zero(space1)  # Identity operator
+    grad_h = odl.ZeroOperator(space1)
+
+    # Check that the algorithm runs. With the above operators, the algorithm
+    # returns the input.
+    x0 = example_element(space1)
+    x = x0.copy()
+    niter = 3
+
+    forward_backward_pd(x, prox_f, prox_cc_g, lin_ops, grad_h, tau=1.0,
+                        sigma=[1.0, 1.0], niter=niter)
+
+    assert x == x0
+
+    # Testing that sizes needs to agree:
+    # Too few sigma_i:s
+    with pytest.raises(ValueError):
+        forward_backward_pd(x, prox_f, prox_cc_g, lin_ops, grad_h, tau=1.0,
+                            sigma=[1.0], niter=niter)
+
+    # Too many operators
+    prox_cc_g_too_many = [odl.solvers.proximal_zero(space1),
+                          odl.solvers.proximal_zero(space1),
+                          odl.solvers.proximal_zero(space1)]
+    with pytest.raises(ValueError):
+        forward_backward_pd(x, prox_f, prox_cc_g_too_many, lin_ops, grad_h,
+                            tau=1.0, sigma=[1.0, 1.0], niter=niter)
+
+    # Test for correct space
+    space2 = odl.uniform_discr(1, 2, 10)
+    x = example_element(space2)
+    with pytest.raises(ValueError):
+        forward_backward_pd(x, prox_f, prox_cc_g, lin_ops, grad_h, tau=1.0,
+                            sigma=[1.0, 1.0], niter=niter)
+
+
+def test_forward_backward_basic():
+    """Test for the forward-backward solver by minimizing ||x||_2^2.
+
+    The general problem is of the form
+
+        ``min_x f(x) + sum_i g_i(L_i x) + h(x)``
+
+    and here we take f(x) = g(x) = 0, h(x) = ||x||_2^2 and L is the
+    zero-operator.
+    """
+
+    space = odl.rn(10)
+
+    lin_ops = [odl.ZeroOperator(space)]
+    prox_cc_g = [odl.solvers.proximal_cconj(odl.solvers.proximal_zero(space))]
+    prox_f = odl.solvers.proximal_zero(space)
+    grad_h = 2.0 * odl.IdentityOperator(space)  # Gradient of two-norm square
+
+    x = example_element(space)
+    x_global_min = space.zero()
+
+    forward_backward_pd(x, prox_f, prox_cc_g, lin_ops, grad_h, tau=0.5,
+                        sigma=[1.0], niter=10)
+
+    assert all_almost_equal(x, x_global_min, places=HIGH_ACCURACY)
+
+
+def test_forward_backward_with_lin_ops():
+    """Test for the forward-backward solver with linear operatros.
```

Comment:
Just a general statement: I really like these introductions, well written, concise and sufficient.